### ISSUE TITLE
Running minishift update gives "Updated successfully" erroneously 

### ISSUE THREAD

**Issue by gastaldi** [minishift/minishift#1090]
Running on Fedora 25, when I run `minishift update` in Minishift 1.1.0 I get the following output:

```
$ minishift update
A newer version of minishift is available.
Do you want to update from 1.1.0 to 1.2.0 now? [y/N]: y
Downloading https://github.com/minishift/minishift/releases/download/v1.2.0/minishift-1.2.0-linux-amd64.tgz
 3.68 MiB / 3.68 MiB [============================================================================================================================================================] 100.00% 0s
 65 B / 65 B [====================================================================================================================================================================] 100.00% 0s

Updated successfully to minishift v1.2.0.

Do you want to update default addons? [y/N]: y
Default add-ons will be updated when executing the next minishift command.
$ minishift status
Minishift was upgraded from v1.1.0 to v1.1.0. Running post update actions.
--- Updating default add-ons ... OK
Default add-ons anyuid, admin-user, xpaas installed
Running
$ minishift version
minishift v1.1.0
```

Notice that when I do `minishift version`, the version is not really updated as the command output says. 

It works if I run it again with `sudo minishift update`

**Comment by hferentschik** [comment]
We had this report once before. I think as well on Linux. Needs some investigation. Not sure why sudo would be required here. How did you install 1.1.0?

**Comment by gastaldi** [comment]
@hferentschik I downloaded the 1.1.0 release manually from https://github.com/minishift/minishift/releases and moved the minishift binary to `/usr/bin` (which requires sudo permission)

**Comment by LalatenduMohanty** [comment]
@gastaldi That explains it.  The binary actually gets replaced so If you have kept the old binary at a place which needs sudo access you need to call it with sudo.

**Comment by LalatenduMohanty** [comment]
I can reproduce this. 

```
$ minishift version
minishift v1.1.0
$ which minishift 
/usr/bin/minishift
 
[lmohanty@LalatenduM-laptop minishift-test-update]A newer version of minishift is available.
Do you want to update from 1.1.0 to 1.2.0 now? [y/N]: y
Downloading https://github.com/minishift/minishift/releases/download/v1.2.0/minishift-1.2.0-linux-amd64.tgz
 3.68 MiB / 3.68 MiB [========================================================================] 100.00% 0s
 65 B / 65 B [========================================================================] 100.00% 0s

Updated successfully to minishift v1.2.0.

$ minishift version
minishift v1.1.0
```

**Comment by hferentschik** [comment]
@LalatenduMohanty, but is there an error? AFAICT there code does handle errors and passes them back. Especially update.Apply which I would think fails.

**Comment by LalatenduMohanty** [comment]
@hferentschik The issue is it does not return error when it fails to replace the binary because of an access issue.

**Comment by gbraad** [comment]
At some point an error should have occurred,  but has been ignored. It
might be worthwhile to check for this. We need to show errors,  or catch
them,  and handle with our own logic... At the moment it shows 'success'
which is not the case

-- 

   Gerard Braad | http://gbraad.nl
   [ Doing Open Source Matters ]

**Comment by hferentschik** [comment]
> At some point an error should have occurred,  but has been ignored.

That I am not so sure about. It might be that go-update does not handle this correctly and is not returning an error. So maybe there is a big in there or maybe we need to add some additional checks of our own.

> It might be worthwhile to check for this. 

Of course

> We need to show errors,  or catch them,  and handle with our own logic... At the moment it shows 'success' which is not the case

+100, this is actually a comment which applies generally. There is a tendency to just check for an error and if it exist pass it on unhanded. This might be ok for many/most cases, but one should at least think about whether better error handling is possible and which actual error is returned. Maybe it contains meta information which helps re-trying an action or taking other recovery actions.

**Comment by gbraad** [comment]
This issue is caused by a `Permission denied` error? It is common for people to move the executable after download to `/usr/bin` or `/usr/local/bin` and they need `sudo` to do this.

**Comment by gastaldi** [comment]
Yes, that's right

**Comment by praveenkumar** [comment]
Looks like go-update doesn't have leaky abstraction but need to do some more check.

```go
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func check() error {
	// get the directory the executable exists in
	TargetPath := "/usr/bin/minishift"
	updateDir := filepath.Dir(TargetPath)
	filename := filepath.Base(TargetPath)

	// Copy the contents of newbinary to a new executable file
	newPath := filepath.Join(updateDir, fmt.Sprintf(".%s.new", filename))
	fp, err := os.OpenFile(newPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0755)
	if err != nil {
		return err
	}
	defer fp.Close()
	return nil
}

func main() {
	err := check()
	fmt.Printf("%#v", err.Error())
}
```

```
$ go run test.go
"open /usr/bin/.minishift.new: permission denied"
```
```

**Comment by hferentschik** [comment]
Merged via pull request #1115
